### ISSUE TITLE
ToolWindowFactory: NPE when opening a project

### ISSUE THREAD

**Issue by lheimbs** [amosproj/amos2023ws02-pitest-ide-plugin#210]
**Describe the bug**
Opening a new project results in a NullPointerException:
```
java.lang.NullPointerException: Cannot invoke "com.intellij.ui.content.Content.getComponent()" because the return value of "com.intellij.ui.content.ContentManager.findContent(String)" is null
	at com.amos.pitmutationmate.pitmutationmate.ui.ToolWindowFactory$Util.updateReport(ToolWindowFactory.kt:62)
	at com.amos.pitmutationmate.pitmutationmate.activity.LaunchActivity.execute$lambda$2$lambda$1(LaunchActivity.kt:34)
	at com.intellij.openapi.application.TransactionGuardImpl.runWithWritingAllowed(TransactionGuardImpl.java:209)
	at com.intellij.openapi.application.TransactionGuardImpl.access$100(TransactionGuardImpl.java:22)
	at com.intellij.openapi.application.TransactionGuardImpl$1.run(TransactionGuardImpl.java:191)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:843)
	at com.intellij.openapi.application.impl.ApplicationImpl$4.run(ApplicationImpl.java:463)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithEnabledImplicitRead(RwLockHolder.kt:75)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithImplicitRead(RwLockHolder.kt:67)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWithImplicitRead(ApplicationImpl.java:1430)
	at com.intellij.openapi.application.impl.FlushQueue.doRun(FlushQueue.java:82)
	at com.intellij.openapi.application.impl.FlushQueue.runNextEvent(FlushQueue.java:124)
	at com.intellij.openapi.application.impl.FlushQueue.flushNow(FlushQueue.java:44)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:318)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:792)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:739)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:733)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:761)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:695)
	at com.intellij.ide.IdeEventQueue._dispatchEvent$lambda$12(IdeEventQueue.kt:589)
	at com.intellij.openapi.application.impl.RwLockHolder.runWithoutImplicitRead(RwLockHolder.kt:44)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:589)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:72)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:355)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:354)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:793)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:354)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:349)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:1014)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:106)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:1014)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$7(IdeEventQueue.kt:349)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:848)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:391)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
```

**To Reproduce**

1. Open a project

**Desktop (please complete the following information):**

- OS: Linux
- IntelliJ IDEA 2023.3.2 (Community Edition)
- Build #IC-233.13135.103, built on December 22, 2023

## Acceptance criteria
- Fix described bug

## Definition of done (DoD)

- All acceptance criteria are met.
- Work products are uploaded to the GitHub repository.
- A pull request is created for each related branch.
- The work products in the pull requests are reviewed.
- GitHub CI Workflow passes for the branches
- The corresponding branches are merged and closed.
- The bill of materials section of the planning documents is updated.
- Tests are written for the added features, if suitable

## DoD general criteria

- Feature has been fully implemented
- Feature has been merged into the mainline
- All acceptance criteria were met
- Product owner approved features
- All tests are passing
- Developers agreed to release

**Comment by nikomall34** [comment]
I also noticed this issue. I could only reproduce it once. Does it occur more frequently for you?

It showed me that the NPE occurs in this line
`
for (file in this.reportDirectory.listFiles()!!) {
`

Maybe the reportDirectory is null? So we have to add a check before that?

**Comment by nikomall34** [comment]
A simple solution for the issue you mentioned would be:
```
val reportWindowContent = toolWindow.contentManager.findContent(PiTestReports.TITLE) ?: return
val reportWindow = reportWindowContent.component
```

**Comment by lheimbs** [comment]
> But I noticed a similiar issue when opening a project in a new window. Unfortunatley I could only reproduce it once.
> 
> It showed me that the NPE occurs in this line of the RunArchiver `for (file in this.reportDirectory.listFiles()!!) {`
> 
> Maybe the reportDirectory is null? So we have to add a check before that?

The `RunArchiver` exception is - as far as i can see - not related to this, maybe create a separate issue.

**Comment by lheimbs** [comment]
> A simple solution for the issue you mentioned would be:
> 
> ```
> val reportWindowContent = toolWindow.contentManager.findContent(PiTestReports.TITLE) ?: return
> val reportWindow = reportWindowContent.component
> ```

I think it is caused by the launch activity launching "too early". Having it run "later" would in my opinion be a better fix and would also solve some other inconsistencies with the plugins-error-message

**Comment by nikomall34** [comment]
Concerning the RunArchiver excpetion you are right it was not related to that. I noticed that it appears whenver the pitest run ended in an excpetion because then no files are created so this.reportDirectory.listFiles() is null. I fixed that in #211.

About the launch activity I agree with you but I thought we are doing that through this line `ApplicationManager.getApplication().invokeLater`

**Comment by lheimbs** [comment]
> I thought we are doing that through this line `ApplicationManager.getApplication().invokeLater`

As i understand it `invokeLater` just makes sure it runs on the AWT thread.
The issue i think is this:
> When the user clicks on the tool window button, the createToolWindowContent() method of the factory class is called, and initializes the UI of the tool window. This procedure ensures that unused tool windows don't cause any overhead in startup time or memory usage: if a user does not interact with the tool window, no plugin code will be loaded or executed.
[Source](https://plugins.jetbrains.com/docs/intellij/tool-windows.html#declarative-setup)

So the tool window has never been created, yet we invoke the update method on it...

**Comment by nikomall34** [comment]
If you have time today we can also discuss it via discord.

**Comment by lheimbs** [comment]
Closing this as it has been resolved in #216
